src/chat/chat_panel.rs

```diff
@@ -594,105 +518,67 @@ impl Widget for ChatPanel {
-                        .downloads
-                        .downloaded_files
-                        .iter()
-                        .find(|file| file.file.id == file_id)
-                        .expect("Attempted to start chat with a no longer existing file")
-                        .clone();
-                    self.load_model(store, downloaded_file);
-                }
-                _ => {}
-            }
-
-            match action.as_widget_action().cast() {
-                ChatLineAction::Delete(id) => {
-                    let store = scope.data.get_mut::<Store>().unwrap();
-                    store.chats.delete_chat_message(id);
-                    self.redraw(cx);
-                }
-                ChatLineAction::Edit(id, updated, regenerate) => {
-                    let store = scope.data.get_mut::<Store>().unwrap();
-                    store.chats.edit_chat_message(id, updated, regenerate);
-
-                    if regenerate {
-                        self.state = ChatPanelState::Streaming {
-                            auto_scroll_pending: true,
-                            auto_scroll_cancellable: false,
-                        };
-
-                        self.show_prompt_input_stop_icon(cx);
-
-                        let prompt_input = self.text_input(id!(main_prompt_input.prompt));
-                        prompt_input.set_text_and_redraw(cx, "");
-                        prompt_input.set_cursor(0, 0);
-                        self.update_prompt_input(cx);
-                    }
-                    self.redraw(cx);
+            } else if let ModelSelectorAction::Selected(downloaded_file) = action.cast() {
+                self.load_model(store, downloaded_file);
+                self.redraw(cx);
+            } else if let ChatAction::Start(file_id) = action.cast() {
+                let downloaded_file = store
+                    .downloads
+                    .downloaded_files
+                    .iter()
+                    .find(|file| file.file.id == file_id)
+                    .expect("Attempted to start chat with a no longer existing file")
+                    .clone();
+                self.load_model(store, downloaded_file);
+            } else if let ChatLineAction::Delete(id) = action.cast() {
+                store.chats.delete_chat_message(id);
+                self.redraw(cx);
+            } else if let ChatLineAction::Edit(id, updated, regenerate) = action.cast() {
+                store.chats.edit_chat_message(id, updated, regenerate);
+                self.redraw(cx);
+            } else if let ChatPanelAction::UnloadIfActive(file_id) = action.cast() {
+                if store
+                    .chats
+                    .get_current_chat()
+                    .map_or(false, |chat| chat.borrow().file_id == file_id)
+                {
+                    self.unload_model(cx);
+                    store.chats.eject_model().expect("Failed to eject model");
                 }
```

Comment:
can we use a match here? we would avoid the else if

Reply:
A single match is not possible because the types are different. This is also why we need to call `.cast()` multiple times.

The previous code did something like

```rust
match action.cast() {
  TypeA::SomeVariant => { ... }
  _ => {}
}

match action.cast() {
  TypeB::SomeVariant => { ... }
  _ => {}
}

match action.cast() {
  TypeC::SomeVariant => { ... }
  _ => {}
}
```

With a lot of `_ => {}` and single branch matches (I think there was one match with two branches, but only one of them)

So... those matches are equivalent to

```rust 
if let TypeA::SomeVariant = action.cast() {
  ...
}

if let TypeB::SomeVariant = action.cast() {
  ...
}

if let TypeC::SomeVariant = action.cast() {
  ...
}
```

Which is the more readable approach for me. And also, [this what clippy would recommend you to do](https://rust-lang.github.io/rust-clippy/master/index.html#/single_match)

I compacted the ifs with `else if` to avoid checking all conditions if one previously matched (as it can only match one of them), but I must agree that `else if` makes the thing look "tight".

Do you agree with not using `else if` but still use `if let`?

Reply:
ooh ok, didn't realize that, its fine then yea